### DISCUSSION TITLE
Networking support?

### DISCUSSION THREAD

**Discussion by mosheduminer** [titzer/virgil#105] [Q&A] [answered]
Hi! I'm poking around at virgil (I find it very intriguing), but I'm wondering about I/O support. It seems to me that there is currently support for file reading/writing, but no support for sockets/networking. Is that correct? How would one go about adding such support (let's say just for Linux)?

**Comment by titzer** [comment, marked as answer]
That's right, there isn't support yet for sockets. To add that for Linux, one could use the kernel system calls directly: https://blog.rchapman.org/posts/Linux_System_Call_Table_for_x86_64/

It's a little tricky to do all of that manually, but is doable and can be wrapped up into a nice library.

**Reply by mosheduminer** [reply]
Thanks. Yeah, I figured I'd need to write syscalls 🤔.

So, it looks like I would need to use `Pointer.atContents` for syscalls that accept pointers. The thing is that I only see how to do this with `Array<byte>` (as documented [here](https://github.com/titzer/virgil/blob/3809902c7ca060ef6965bd7dbf793b79c1d9a21b/doc/tutorial/Pointers.md)), but I'm not sure what to do about syscalls that accept a struct, like `sockaddr`?

**Reply by titzer** [reply]
I usually use an `Array<long>` or `Array<int>` and poke the individual words, e.g. see the use of `stat` in `rt/x86-64-linux/System.v3`. I'm working on a new language construct to allow defining structs with known layouts (which are really just views over byte arrays). Hopefully that will be ready from prime time soon. (see `test/struct/` for what that will look like).

**Reply by mosheduminer** [reply]
I see, so in the meantime I can basically just implement structs as over byte arrays. Thanks 👍. I'll post back here if I make substantial progress.

**Reply by titzer** [reply]
Great. Also note that you can do pointer arithmetic and also load/store to a `Pointer` in quantities more than a byte, like:

```
var data: Array<byte>;
var ptr = Pointer.atContents(data);
(ptr + 16).store<u64>(expr);
...
```

**Reply by mosheduminer** [reply]
Hey, I'm trying to figure out how to interpret the returned values frol syscalls. I expect both the socket and connect syscalls to return an int, but in virgil they both return a tuple of two ints. Mind helping me out here? 

I did try (and fail) to figure out where in the repo syscalls are implemented.

**Comment by titzer** [comment]
Are you targeting `x86-64-linux`? IIRC, the first element of the tuple is the return value and the second is the errno. See for example `rt/x86-64-linux/System.v3`. The `Linux.syscall` is implemented in the compiler in `aeneas/src/x86-64/SsaX86_64Gen.v3`. The return values are just the registers set by the kernel, so `rax` and `rdx`.

**Reply by mosheduminer** [reply]
Thanks. I think I'm missing something here, because the connect call (to a local, closed port) returns the tuple `(-111, 16)`, but based on the manpage, I'm expecting the first value to be `-1`. Also, I've discovered that the second value is just whatever value is passed in as `addrlen`...

I'm gonna take a break from this for now.

(I have little experience with network / systems programming in general).

**Reply by mosheduminer** [reply]
I've decided I'm not going to be able to get this working until I can just use structs. I'm pretty sure that there's something wrong with how I'm encoding the data, but I'm not sure what. So I've shelved this project for now.

**Reply by titzer** [reply]
Ok, no worries. It's admittedly a real pain to do it with pointer arithmetic. I have a student working on the `struct` implementation, so hopefully this will work in ~1-2 months.

**Reply by mosheduminer** [reply]
Alright 👍. I suppose I'll try again then.

**Comment by titzer** [comment]
Random debugging tip: Pass `-symbols` as a command-line option to `v3c` and it will generate symbols with function boundaries, allowing you to step through the asm in `gdb` and use breakpoints. So you can at least debug at the assembly level.

It's low-level, I know. Debugging support could be improved, for sure. But symbols were a lifesaver for me.

**Comment by mosheduminer** [comment]
I just noticed https://github.com/titzer/virgil/commit/8520fa83091ca80de3b49b0a89b27183a2abfc19! Very exciting.

**Comment by mosheduminer** [comment]
Since @k-sareen is writing network code from scratch, I thought I'd share this here (number 1 on HN right now) https://news.ycombinator.com/item?id=34179426.

**Reply by k-sareen** [reply]
Interesting read! Thanks for the link. Currently there's no support for socket options, but they should be exposed to the user. Don't know what is the best way to do that. Maybe just a `flags` field which is passed into the socket constructor?

**Reply by mosheduminer** [reply]
I guess so. It would be nice to maybe have keyword arguments for each flag but we don't have those in virgil I think.

**Reply by k-sareen** [reply]
Right. It's super barebones right now. I'll add the constant definitions to `LinuxConst`.
